tsdb/engine/tsm1/engine_test.go

```diff
@@ -250,6 +251,274 @@ func TestEngine_Backup(t *testing.T) {
 	}
 }
 
+func TestEngine_Export(t *testing.T) {
+	// Generate temporary file.
+	f, _ := ioutil.TempFile("", "tsm")
+	f.Close()
+	os.Remove(f.Name())
+	walPath := filepath.Join(f.Name(), "wal")
+	os.MkdirAll(walPath, 0777)
+	defer os.RemoveAll(f.Name())
+
+	// Create a few points.
+	p1 := MustParsePointString("cpu,host=A value=1.1 1000000000")
+	p2 := MustParsePointString("cpu,host=B value=1.2 2000000000")
+	p3 := MustParsePointString("cpu,host=C value=1.3 3000000000")
+
+	// Write those points to the engine.
+	db := path.Base(f.Name())
+	opt := tsdb.NewEngineOptions()
+	opt.InmemIndex = inmem.NewIndex(db)
+	idx := tsdb.MustOpenIndex(1, db, filepath.Join(f.Name(), "index"), opt)
+	defer idx.Close()
+
+	e := tsm1.NewEngine(1, idx, db, f.Name(), walPath, opt).(*tsm1.Engine)
+
+	// mock the planner so compactions don't run during the test
+	e.CompactionPlan = &mockPlanner{}
+
+	if err := e.Open(); err != nil {
+		t.Fatalf("failed to open tsm1 engine: %s", err.Error())
+	}
+
+	if err := e.WritePoints([]models.Point{p1}); err != nil {
+		t.Fatalf("failed to write points: %s", err.Error())
+	}
+	if err := e.WriteSnapshot(); err != nil {
+		t.Fatalf("failed to snapshot: %s", err.Error())
+	}
+
+	if err := e.WritePoints([]models.Point{p2}); err != nil {
+		t.Fatalf("failed to write points: %s", err.Error())
+	}
+	if err := e.WriteSnapshot(); err != nil {
+		t.Fatalf("failed to snapshot: %s", err.Error())
+	}
+
+	if err := e.WritePoints([]models.Point{p3}); err != nil {
+		t.Fatalf("failed to write points: %s", err.Error())
+	}
+
+	// export the whole DB
+	var exBuf bytes.Buffer
+	if err := e.Export(&exBuf, "", time.Unix(0, 0), time.Unix(0, 4000000000)); err != nil {
+		t.Fatalf("failed to export: %s", err.Error())
+	}
+
+	var bkBuf bytes.Buffer
+	if err := e.Backup(&bkBuf, "", time.Unix(0, 0)); err != nil {
+		t.Fatalf("failed to backup: %s", err.Error())
+	}
+
+	if len(e.FileStore.Files()) != 3 {
+		t.Fatalf("file count wrong: exp: %d, got: %d", 3, len(e.FileStore.Files()))
+	}
+
+	fileNames := map[string]bool{}
+	for _, f := range e.FileStore.Files() {
+		fileNames[filepath.Base(f.Path())] = true
+	}
+
+	fileData, err := getExportData(&exBuf)
+	if err != nil {
+		t.Errorf("Error extracting data from export: %s", err.Error())
+	}
+
+	// TEST 1: did we get any extra files not found in the store?
+	for k, _ := range fileData {
+		if _, ok := fileNames[k]; !ok {
+			t.Errorf("exported a file not in the store: %s", k)
+		}
+	}
+
+	// TEST 2: did we miss any files that the store had?
+	for k, _ := range fileNames {
+		if _, ok := fileData[k]; !ok {
+			t.Errorf("failed to export a file from the store: %s", k)
+		}
+	}
+
+	// TEST 3: Does 'backup' get the same files + bits?
+	tr := tar.NewReader(&bkBuf)
+
+	th, err := tr.Next()
+	for err == nil {
+		expData, ok := fileData[th.Name]
+		if !ok {
+			t.Errorf("Extra file in backup: %q", th.Name)
+			continue
+		}
+
+		buf := new(bytes.Buffer)
```

Comment:
tar readers know the size of the files. You could replace this with a fixed-size buffer and use `ReadFull` rather than having to use a `bytes.Buffer` everywhere.